File: shared/engines/mediapipe_engine.py
from typing import Optional, Dict, Any, List
from shared.engines.interfaces import MediaPipeEngineInterface
from shared.events.event_bus import EventBus
from shared.types.enums import EngineStatus
from shared.contracts.vision import Landmark, LandmarkSet


class MediaPipeEngine(MediaPipeEngineInterface):
    """Python MediaPipe Engine interface implementation for PostureSense server-side Engine Runtime."""

    def __init__(self, name: str = "MediaPipeEngine", event_bus: Optional[EventBus] = None):
        super().__init__(name=name, event_bus=event_bus)
        self.version = "2.0.0"
        self.priority = 2
        self.dependencies = ["camera_engine"]
        self._is_model_loaded = False
        self._is_tracking = False
        self._frames_processed = 0
        self._inference_latency_ms = 0.0

    def initialize(self, config: Optional[dict] = None) -> bool:
        self._status = EngineStatus.INITIALIZED
        self.publish("mediapipe.initialized", self.get_diagnostics())
        self.load_model()
        return True

    def load_model(self) -> bool:
        self._is_model_loaded = True
        self.publish("mediapipe.model_loaded", self.get_diagnostics())
        return True

    def start(self) -> bool:
        self._status = EngineStatus.RUNNING
        self.publish("mediapipe.started", self.get_diagnostics())
        return True

    def pause(self) -> bool:
        self._status = EngineStatus.PAUSED
        self.publish("mediapipe.paused", self.get_diagnostics())
        return True

    def resume(self) -> bool:
        self._status = EngineStatus.RUNNING
        self.publish("mediapipe.resumed", self.get_diagnostics())
        return True

    def stop(self) -> bool:
        self._status = EngineStatus.STOPPED
        self.publish("mediapipe.stopped", self.get_diagnostics())
        return True

    def dispose(self) -> bool:
        self._status = EngineStatus.DISPOSED
        self.publish("mediapipe.disposed", self.get_diagnostics())
        return True

    def process_raw_frame(self, frame_number: int = 1) -> LandmarkSet:
        self._frames_processed += 1
        self._inference_latency_ms = 12.5

        # 33 raw MediaPipe pose keypoints
        landmarks = [
            Landmark(index=i, name=f"KEYPOINT_{i}", x=0.5, y=0.5, z=0.0, visibility=0.99)
            for i in range(33)
        ]
        landmark_set = LandmarkSet(landmarks=landmarks, confidence=0.95, source=self.name)

        if not self._is_tracking:
            self._is_tracking = True
            self.publish("tracking.recovered", {"confidence": 0.95})

        self.publish("landmarks.detected", landmark_set.to_dict())
        return landmark_set

    def get_diagnostics(self) -> Dict[str, Any]:
        return {
            "name": self.name,
            "version": self.version,
            "status": self._status.value if hasattr(self._status, 'value') else str(self._status),
            "priority": self.priority,
            "dependencies": self.dependencies,
            "is_model_loaded": self._is_model_loaded,
            "is_tracking": self._is_tracking,
            "metrics": {
                "frames_processed": self._frames_processed,
                "inference_latency_ms": self._inference_latency_ms,
                "landmark_count": 33
            }
        }

File: static/assets/js/engines/mediapipe_engine.js
/**
 * MediaPipeEngine
 * Production-grade client MediaPipe Tasks Vision Pose Landmarker wrapper engine for PostureSense v2.
 * Offloads inference to a Web Worker, processes camera frames, emits landmarks.detected & tracking.lost events,
 * and handles model load failures gracefully.
 */

export class MediaPipeEngine {
    constructor(eventBus = null) {
        this.name = "MediaPipeEngine";
        this.version = "2.0.0";
        this.eventBus = eventBus;
        this.status = "uninitialized";
        this.priority = 2;
        this.dependencies = ["camera_engine"];

        this.worker = null;
        this.isModelLoaded = false;
        this.isTracking = false;
        this.isInferenceBusy = false;
        this.fpsTimer = null;
        this.frameCountInSecond = 0;

        // Metrics & Diagnostics
        this.metrics = {
            modelLoadTimeMs: 0.0,
            inferenceLatencyMs: 0.0,
            inferenceFps: 0.0,
            framesReceived: 0,
            framesProcessed: 0,
            framesDropped: 0,
            landmarkCount: 0,
            trackingConfidence: 0.0
        };
    }

    async initialize(config = {}) {
        this.status = "initialized";
        this._publish("mediapipe.initialized", this.getDiagnostics());
        return true;
    }

    async loadModel() {
        const startTime = performance.now();
        console.log('[MediaPipeEngine] Creating MediaPipe module worker...');

        return new Promise((resolve) => {
            try {
                this.worker = new Worker('/static/assets/js/workers/mediapipe_worker.js', { type: 'module' });

                this.worker.onerror = (err) => {
                    const errorMsg = err.message || 'MediaPipe Web Worker failed to load or execute.';
                    console.error('[MediaPipeEngine] Worker error:', errorMsg);
                    this.status = 'failed';
                    this.isModelLoaded = false;
                    this._publish("mediapipe.failed", { error: errorMsg });
                    this._publish("tracking.lost", {
                        reason: "MediaPipe Worker Error",
                        error: errorMsg,
                        timestamp: new Date().toISOString()
                    });
                    resolve(false);
                };

                this.worker.onmessage = (e) => {
                    const { action, success, landmarks, confidence, latencyMs, error } = e.data;
                    if (action === 'MODEL_LOADED' && success) {
                        this.isModelLoaded = true;
                        this.metrics.modelLoadTimeMs = performance.now() - startTime;
                        this.status = 'ready';
                        this._publish("mediapipe.model_loaded", this.getDiagnostics());
                        resolve(true);
                    } else if (action === 'MODEL_ERROR') {
                        this.status = "failed";
                        this.isModelLoaded = false;
                        this._publish("mediapipe.failed", { error });
                        this._publish("tracking.lost", {
                            reason: "MediaPipe Model Load Failure",
                            error: error,
                            timestamp: new Date().toISOString()
                        });
                        resolve(false);
                    } else if (action === 'FRAME_PROCESSED') {
                        this.isInferenceBusy = false;
                        this._handleFrameProcessed(landmarks || [], confidence || 0.0, latencyMs || 0.0);
                    } else if (action === 'FRAME_ERROR') {
                        this.isInferenceBusy = false;
                        console.warn('[MediaPipeEngine] Frame processing error:', error);
                        this._publish("mediapipe.error", { error });
                        this._publish("tracking.lost", {
                            reason: "MediaPipe Frame Error",
                            error: error,
                            timestamp: new Date().toISOString()
                        });
                    }
                };

                this.worker.postMessage({ action: 'LOAD_MODEL' });
            } catch (err) {
                const errorMsg = err.message || String(err);
                console.warn('[MediaPipeEngine] Web Worker init exception:', errorMsg);
                this.status = "failed";
                this.isModelLoaded = false;
                this._publish("mediapipe.failed", { error: errorMsg });
                this._publish("tracking.lost", {
                    reason: "MediaPipe Web Worker Init Exception",
                    error: errorMsg,
                    timestamp: new Date().toISOString()
                });
                resolve(false);
            }
        });
    }

    async warmup() {
        if (this.isModelLoaded) {
            console.log('[MediaPipeEngine] Model warmed up.');
        }
    }

    async start() {
        if (!this.isModelLoaded) {
            const loaded = await this.loadModel();
            if (!loaded) {
                console.error('[MediaPipeEngine] Start aborted — model loading failed.');
                this.status = 'failed';
                return false;
            }
        }
        this.status = "running";
        this.isInferenceBusy = false;
        this._startFpsTimer();
        this._subscribeToCameraFrames();
        this._publish("mediapipe.started", this.getDiagnostics());
        return true;
    }

    pause() {
        if (this.status === 'running') {
            this.status = "paused";
            this.isInferenceBusy = false;
            this._publish("mediapipe.paused", this.getDiagnostics());
        }
    }

    resume() {
        if (this.status === 'paused') {
            this.status = "running";
            this.isInferenceBusy = false;
            this._publish("mediapipe.resumed", this.getDiagnostics());
        }
    }

    async stop() {
        this.status = "stopped";
        this.isInferenceBusy = false;
        if (this.fpsTimer) {
            clearInterval(this.fpsTimer);
            this.fpsTimer = null;
        }
        this._publish("mediapipe.stopped", this.getDiagnostics());
    }

    dispose() {
        this.stop();
        if (this.worker) {
            this.worker.terminate();
            this.worker = null;
        }
        this.status = "disposed";
        this._publish("mediapipe.disposed", this.getDiagnostics());
    }

    _subscribeToCameraFrames() {
        if (!this.eventBus || typeof this.eventBus.subscribe !== 'function') return;

        const onFrame = async (event) => {
            if (this.status !== 'running' || !this.worker) return;

            this.metrics.framesReceived++;
            if (this.metrics.framesReceived % 30 === 0) {
                const data = event.data || {};
                console.log(`[MediaPipeEngine] Frames received: ${this.metrics.framesReceived}`, {
                    frameNumber: data.frame_number || data.frameNumber || this.metrics.framesReceived,
                    width: data.width || 1280,
                    height: data.height || 720,
                    timestamp: data.timestamp || new Date().toISOString()
                });
            }

            if (this.isInferenceBusy) {
                this.metrics.framesDropped++;
                if (event.data && event.data.imageBitmap && typeof event.data.imageBitmap.close === 'function') {
                    try { event.data.imageBitmap.close(); } catch (_) {}
                }
                return;
            }

            let bitmap = event.data?.imageBitmap || null;
            if (!bitmap && event.data?.videoElement && event.data.videoElement.readyState >= 2) {
                try {
                    bitmap = await createImageBitmap(event.data.videoElement);
                } catch (_) {}
            }

            if (bitmap) {
                this.isInferenceBusy = true;
                this.metrics.framesProcessed++;
                this.frameCountInSecond++;
                this.worker.postMessage({
                    action: 'PROCESS_FRAME',
                    payload: {
                        imageBitmap: bitmap,
                        frameNumber: this.metrics.framesProcessed,
                        timestamp: performance.now()
                    }
                }, [bitmap]);
            }
        };

        this.eventBus.subscribe('frame.captured', onFrame);
        this.eventBus.subscribe('camera.frame_ready', onFrame);
    }

    _handleFrameProcessed(landmarks, confidence, latencyMs) {
        this.metrics.inferenceLatencyMs = latencyMs || 0.0;
        this.metrics.landmarkCount = landmarks.length;
        this.metrics.trackingConfidence = confidence || 0.0;

        if (this.metrics.framesProcessed % 30 === 0) {
            console.log('[MediaPipeEngine] landmarks.detected:', {
                frameNumber: this.metrics.framesProcessed,
                landmarkCount: landmarks.length,
                confidence: confidence
            });
        }

        if (landmarks.length > 0) {
            if (!this.isTracking) {
                this.isTracking = true;
                this._publish("tracking.recovered", { confidence });
            }

            const landmarkSetPayload = {
                id: Math.random().toString(36).substring(2, 11),
                timestamp: new Date().toISOString(),
                schema_version: '2.0.0',
                source: 'MediaPipeEngine',
                confidence: confidence,
                landmarks: landmarks
            };

            this._publish("landmarks.detected", landmarkSetPayload);
        } else {
            if (this.isTracking) {
                this.isTracking = false;
                this._publish("tracking.lost", {
                    reason: "No landmarks detected in frame",
                    confidence: 0.0,
                    timestamp: new Date().toISOString()
                });
            }
        }
    }

    _startFpsTimer() {
        this.fpsTimer = setInterval(() => {
            this.metrics.inferenceFps = this.frameCountInSecond;
            this.frameCountInSecond = 0;
        }, 1000);
    }

    getDiagnostics() {
        return {
            name: this.name,
            version: this.version,
            status: this.status,
            priority: this.priority,
            dependencies: this.dependencies,
            config: {
                isModelLoaded: this.isModelLoaded,
                isTracking: this.isTracking
            },
            metrics: { ...this.metrics }
        };
    }

    _publish(eventName, data) {
        if (this.eventBus && typeof this.eventBus.publish === 'function') {
            this.eventBus.publish(eventName, data);
        }
    }
}
